SemanBoon/SheBraids
# User Preference of Factors and Readjustment of Weights in the Utility Function

## What type of PR is this?



## Description

This PR introduces a new feature to the booking page that allows users to specify their priority factor when booking a service. The available priority factors are:

- Focus Block (appointments immediately before or after another appointment)
- User Preferred Time (based on past booking trends)
- Popular Slots (popular times for the service provider)

 Based on the selected priority, the weights in the utility function of the recommendation algorithm are adjusted to provide more relevant time slot recommendations.

## Screenshots, Recordings

<img width="785" alt="Screenshot 2024-07-31 at 11 26 44 AM" src="https://github.com/user-attachments/assets/72e4f1f3-5e9d-41e7-ab1c-56326a4333fb">


<img width="732" alt="Screenshot 2024-07-31 at 11 27 04 AM" src="https://github.com/user-attachments/assets/698f380e-f6b3-4ea9-bca7-7f967b4149ec">

<img width="903" alt="Screenshot 2024-07-31 at 11 27 11 AM" src="https://github.com/user-attachments/assets/0b2e409f-d00c-4824-8f96-36ab012e5fcf">
<img width="863" alt="Screenshot 2024-07-31 at 11 27 19 AM" src="https://github.com/user-attachments/assets/82e8dc4e-2bb8-4c99-a444-b0b809eafdc6">

## Test Plan

Manual Testing:
- Go to the booking page.
- Pick a service from the dropdown menu.
- Select a priority factor from the new dropdown.
- Made sure that the recommended slots update based on what I selected.

Unit Tests: Added unit tests to verify that the utility function adjusts weights correctly based on the selected priority.
